src/cfnlint/rules/resources/NoEcho.py

```diff
@@ -0,0 +1,52 @@
+"""
+Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: MIT-0
+"""
+from ast import literal_eval
+from cfnlint.rules import CloudFormationLintRule
+from cfnlint.rules import RuleMatch
+
+
+class NoEcho(CloudFormationLintRule):
+    id = 'W4002'
+    shortdesc = 'Check for NoEcho References'
+    description = 'Check if there is a NoEcho enabled parameter referenced within a resources Metadata section'
+    source_url = 'https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html#parameters-section-structure-properties'
+    tags = ['resources', 'NoEcho']
+
+    def match(self, cfn):
+        matches = []
+        no_echo_params = []
+        parameters = cfn.get_parameters()
+
+        for parameter_name, parameter_value in parameters.items():
+            parameter_properties = literal_eval(str(parameter_value))
+            for parameter_properties_name, parameter_properties_value in parameter_properties.items():
+                if parameter_properties_name == 'NoEcho' and parameter_properties_value is True:
+                    no_echo_params.append(parameter_name)
+
+        resource_properties = cfn.get_resources()
+        for resource_name, resource_values in resource_properties.items():
+
+            resource = get_dict(resource_values)
+            for key, value in resource.items():
+                if key == 'Metadata':
+                    metadata = get_dict(value)
+                    for prop_name, prop_value in metadata.items():
```

Comment:
I don't think `literal_eval(str(input_string))` handles the logic the value isn't a dict already.  If the value is `1` then this will fail (I believe) with `'int' object has no attribute 'items'`.

Reply:
```
Metadata:
      Test: "test"
      NoEchoParamInMetadata: !Ref NoEchoParam
```
`E0002 Unknown exception while processing rule W4002: 'int' object has no attribute 'values'`

Reply:
That's a great point there. I took another look into how the logic was done and found room for improvement, changed in the latest commit!